### PR TITLE
feat (coupons): add possibility to define if coupon will be reusable

### PR DESCRIPTION
**Pull request by lovrocolic**
## Context

Allow reusing coupons multiple times

## Description

Allow defining if certain coupon can be reusable or not

### CODE REVIEW

File: spec/graphql/mutations/coupons/update_spec.rb [hunk 1]
```diff
@@ -51,6 +53,7 @@
       expect(result_data['amountCurrency']).to eq('USD')
       expect(result_data['expiration']).to eq('time_limit')
       expect(result_data['expirationDate']).to eq (Time.current + 33.days).to_date.to_s
+      expect(result_data['reusable']).to eq (false)
```
Comment (jdenquin) [inline spec/graphql/mutations/coupons/update_spec.rb:56]: Lint/ParenthesesAsGroupedExpression: `(false)` interpreted as grouped expression. (https://rubystyle.guide#parens-no-spaces)

File: spec/services/applied_coupons/create_service_spec.rb [hunk 2]
```diff
@@ -162,6 +162,22 @@
...
+        aggregate_failures do
+          expect(create_result).not_to be_success
+          expect(create_result.error).to be_a(BaseService::ValidationFailure)
+          expect(create_result.error.messages.keys).to include(:coupon)
+          expect(create_result.error.messages[:coupon]).to include('coupon_is_not_reusable')
+        end
+      end
+    end
+
+
```
Comment (jdenquin) [inline spec/services/applied_coupons/create_service_spec.rb:180]: Layout/EmptyLines: Extra blank line detected. (https://rubystyle.guide#two-or-more-empty-lines)

**Review by vincent-pochet — APPROVED** [review]
Small comments, but looks good!

File: app/services/applied_coupons/create_service.rb [hunk 3]
```diff
@@ -87,6 +90,13 @@ def process_creation(applied_coupon_attributes)
       result.record_validation_failure!(record: e.record)
     end
 
+    def reusable_coupon?
+      return true if coupon.reusable?
+      return false if customer.applied_coupons.where(coupon_id: coupon.id).exists?
+
+      true
```
[the remark this answers is not in the document]
Comment (lovrocolic) [reply app/services/applied_coupons/create_service.rb]: fixed

**Pull request merged**